### DISCUSSION TITLE
Trackball STEP files cannot be loaded - Docker install issue, or am I doing something wrong?

### DISCUSSION THREAD

**Discussion by elliottmckee** [joshreve/dactyl-keyboard#67] [Q&A] [answered]
Hello!

This is my first attempt at installing and playing around with this, and I am pretty new to Docker, so there is a good chance that I am doing something really simple wrong with regard to either of those, which is causing my issue. I don't know what current Python version I have installed on my machines are. If it is critical, let me know, but I don't know enough about Docker to know if this is the case. 

Basically, I am just trying to experiment with trackball layouts. The only thing I have changed about _run_config.json_ is that I have changed the 'thumb_style' field. Everything seems to work fine, and outputs a reasonable looking .STEP for the non-trackball cases, but when I go to run it with any of the TRACKBALL 'thumb_styles', it throws an error when trying to import the TRACKBALL STEP files. **Again, I am using the perfectly default/clean run_config.json file, with ONLY the 'thumb_styles' field altered.** 

```
...(clipped)...
>>> Traceback (most recent call last):
  File "dactyl_manuform.py", line 4249, in <module>
    run()
  File "dactyl_manuform.py", line 4202, in run
    mod_r, tmb_r = model_side(side="right")
  File "dactyl_manuform.py", line 4023, in model_side
    tbprecut, tb, tbcutout, sensor, ball = generate_trackball_in_cluster()
  File "dactyl_manuform.py", line 3316, in generate_trackball_in_cluster
    return generate_trackball(pos, rot)
  File "dactyl_manuform.py", line 3278, in generate_trackball
    shape, cutout, sensor = trackball_socket()
  File "dactyl_manuform.py", line 317, in trackball_socket
    shape = import_file(tb_file)
  File "/app/src/helpers_cadquery.py", line 227, in import_file
    fname + ".step"))
  File "/opt/conda/envs/myenv/lib/python3.6/site-packages/cadquery/occ_impl/importers.py", line 45, in importShape
    return importStep(fileName, *args, **kwargs)
  File "/opt/conda/envs/myenv/lib/python3.6/site-packages/cadquery/occ_impl/importers.py", line 64, in importStep
    raise ValueError("STEP File could not be loaded")
ValueError: STEP File could not be loaded
```

I will include my problematic config file attached below. 

I am currently using a clean clone of the main Github branch (should I be using the trackball branch?), and using the Docker installation. Specifically, here's what I am doing in terms installation:
1. Clone Github Repo (Main Branch)
2. Run build_docker.bat
3. Modify run_config,json
4. Open Docker, check that DM-shell is running, then run DM-run

What I have tried
- I have tried this on two separate computers, both with two clean clones of the Repo
- I have tried running Docker as an administrator
- I have played around with fiddling with other fields in the Run_Config.json file, to no avail as I only know what a few of them do. _(if anyone has any resources on where to find info on what each of the fields changes, that would be great! learning all this has proven somewhat daunting)_

Any help on this front would be much appreciated. I get the sense that this may not be a problem with the program, and more of user error.  It also could be an issue stemming from the Docker installation...

Let me know if anything else would be useful, and thanks in advance!

Run_Config.json (converted to .txt):
[run_config.txt](https://github.com/joshreve/dactyl-keyboard/files/7611521/run_config.txt)

**Comment by Looooopy** [comment]
Hi @elliottmckee 

I have now tested trackball and for me it looks like it works. 
> UPDATE: Found the symptom you are using `"ENGINE": "cadquery",` switch to `"ENGINE": "solid",` and you should get an output for now.

I can give you my workflow and see if it helps...

**Pre requisites** 
* Local machine
  * Docker
  * OpenSCAD and FreeCAD (use OpenSCAD for testing because its faster) 
  * I vaguely remembered I installed something more... but it might been when I tried out some other branch of dactyl Manuform

**My Workflow**
* Open `generate_configuration.py` in an text editor change engine to `'ENGINE': 'solid'` to produce OpenSCAD files (leave it open)
* Start OpenSCAD
* Open terminal
* Run `./dactyl.sh --build`
* Start DM-shell
* Run this command that generates the `run_config.json` and start producing the output files to `./things` folder
```bash
  python ./generate_configuration.py && python ./dactyl_manuform.py 
```
* Open for example 'XXXleft.scad' in OpenSCAD (leave it open)
* Iterate
  *  Do changes I would like to test in `generate_configuration.py`
  *  Rerun the command above 
  *  Switch fast to OpenSCAD to so you see what changes - if you are not sure :)
  * Keep iterate until you are satisfied with the result
* Use either FreeCAD or OpenSCAD to generate 3D Printer format
  * OpenSCAD 
    * Press F6 to generate the real model
    * Select export from menu
    * Export to `*.stl `or other format for your 3D printer
  * FreeCAD
    * Change to ```'ENGINE': 'cadquery'``` in ```generate_configuration.py```
    * Rerun the command above to produce `*.step` files that you need to open with FreeCAD
    * Export to `*.stl `or other format for your 3D printer

**Reply by joshreve** [reply]
I think I found the issue.  Let me check it and I should have it pushed today.  I think the batch file didn't get the parts folder added to the docker directory mounting.  It should be a fast fix.

**Reply by joshreve** [reply]
The method suggested by @Looooopy should work in the meantime, but I think I'll have the fix very shortly.

**Reply by elliottmckee** [reply]
Hi @Looooopy!

Appreciate the response! I realized I actually want to go with the TRACKBALL_WILD, which appears to still be in an open PR before it is fully merged... at least I think- i'm not the most well-versed in Git. In this process, I also just switched over to running through Anaconda, and that removes the Docker variable that I am not really familiar with. 

As such, I haven't been able to try the above workflow, or check if @joshreve's fix has worked for me. However, it appears that @joshreve has fixed the CADQUERY implementation :)

I do appreciate this description of your workflow- I know there is a high-level description of it in the README, but this was still one of the harder things to pickup and figure out. This description will be useful to me once I end up switching back the the Docker implementation, and hope it will be for others!

**Comment by joshreve** [comment, marked as answer]
Should be fixed in the latest push. Was exactly what I thought. Let me know if you still have issues, but basically the docker container didn't have access to the folder with parts in it and I corrected that.

**Reply by elliottmckee** [reply]
Glad to hear!

I moved on to running the TRACKBALL_WILD PR version on Anaconda, **but I did go back and confirm that this did solve my/the issue of using the TRACKBALL files with CADQUERY while using the Docker implementation.** 

Appreciate the really fast response and resolution; along with everything else you do here! I've just started playing around with all this  and it is all SUPER cool and powerful :)

**Reply by joshreve** [reply]
No problem good to see people using it.  If you are building the Wilder Trackball version, working from his PR / Repo is the best option at the moment.  I am rolling the his improvements into an ongoing functional refactor, but it is too early/unstable to use at the moment due to the scale of the structural refactoring.
